sw/host/opentitantool/src/command/spi.rs

```diff
@@ -0,0 +1,276 @@
+            .template("[{elapsed_precise}] [{wide_bar}] {bytes}/{total_bytes} ({eta})"),
+    );
+    progress
+}
+
+impl CommandDispatch for SpiSfdp {
+    fn run(&self, transport: &mut dyn Transport) -> Result<Option<Box<dyn Serialize>>> {
+        transport.capabilities().request(Capability::SPI).ok()?;
+        let spi = transport.spi()?;
+
+        if let Some(length) = self.raw {
+            let mut buffer = vec![0u8; length];
+            spi.run_transaction(&mut [
+                Transfer::Write(&[0x5a, 0, 0, 0, 0]),
+                Transfer::Read(&mut buffer),
+            ])?;
+            hexdump(&buffer);
+            Ok(None)
+        } else {
+            let sfdp = SpiFlash::read_sfdp(&*spi)?;
+            Ok(Some(Box::new(sfdp)))
+        }
+    }
+}
+
+/// Read the JEDEC ID of a SPI EEPROM.
+#[derive(Debug, StructOpt)]
+pub struct SpiReadId {
+    #[structopt(
+        short = "n",
+        long,
+        default_value = "3",
+        help = "Number of JEDEC ID bytes to read."
+    )]
+    length: usize,
+}
+
+#[derive(Debug, serde::Serialize)]
+pub struct SpiReadIdResponse {
+    jedec_id: Vec<u8>,
+}
+
+impl CommandDispatch for SpiReadId {
+    fn run(&self, transport: &mut dyn Transport) -> Result<Option<Box<dyn Serialize>>> {
+        transport.capabilities().request(Capability::SPI).ok()?;
+        let spi = transport.spi()?;
+        let jedec_id = SpiFlash::read_jedec_id(&*spi, self.length)?;
+        Ok(Some(Box::new(SpiReadIdResponse { jedec_id })))
+    }
+}
+
+/// Read data from a SPI EEPROM.
+#[derive(Debug, StructOpt)]
+pub struct SpiRead {
+    #[structopt(short, long, default_value = "0", help = "Start offset.")]
+    start: u32,
+    #[structopt(
+        short = "n",
+        long,
+        default_value = "4096",
+        help = "Number of bytes to read."
+    )]
+    length: usize,
+    #[structopt(short, long, help = "Hexdump to stdout.")]
+    hexdump: bool,
+    #[structopt(name = "FILE", default_value = "-")]
+    filename: String,
+}
+
+#[derive(Debug, serde::Serialize)]
+pub struct SpiReadResponse {
+    length: usize,
+    bytes_per_second: f64,
+}
+
+impl CommandDispatch for SpiRead {
+    fn run(&self, transport: &mut dyn Transport) -> Result<Option<Box<dyn Serialize>>> {
+        transport.capabilities().request(Capability::SPI).ok()?;
+        let spi = transport.spi()?;
+        let mut flash = SpiFlash::from_spi(&*spi)?;
+        flash.set_address_mode_auto(&*spi)?;
+
+        let mut buffer = vec![0u8; self.length];
+        let progress = progress_bar(self.length as u64);
+        let t0 = Instant::now();
+        flash.read_with_progress(&*spi, self.start, &mut buffer, |_, chunk| {
+            progress.inc(chunk as u64);
+        })?;
+        progress.finish();
+        let t1 = Instant::now();
+        let duration = (t1 - t0).as_secs_f64();
+        let bytes_per_sec = buffer.len() as f64 / duration;
+
+        if self.filename == "-" {
+            if self.hexdump {
```

Comment:
Do you want hexdump to work irrespective of which file is used?

Reply:
Originally, I'd only intended for hexdump to be valid with stdout, but there's really no reason for that restriction.  Fixed to hexdump either to stdout or to the chosen file.